platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,826 @@
+
+Option 2 was chosen for existing networks and Option 8 was selected for new networks. This decision was made to
+ensure that the first block in the block stream is signed by the ledger and that new networks always have block 0
+signed.
+
+---
+
+## Changes
+
+The changes are presented in the following order:
+
+1. Public API
+2. Core Behaviors
+3. Component Architecture
+4. Configuration
+5. Metrics
+6. Performance
+
+### Public API
+
+The public API changes consist of the following:
+
+* The Ledger Signature API
+* System Transactions
+* State Data Structures
+
+#### Ledger Signature API
+
+The Ledger Signature API is the interface between the application and the platform by which the application submits
+a message to be signed by the ledger, the nodes sign the message with their shares, and the platform aggregates the
+signatures into a ledger signature. The ledger signature is then asynchronously returned to the application.
+
+To request a ledger signature, the application invokes a new method on the `State`:
+
+* `State::sign(byte[] message) throws NotReadyException`
+
+To receive a ledger signature, the application registers a callback with the platform builder:
+
+* `PlatformBuilder::withLedgerSignatureCallback(Consumer<PairingSignature> ledgerSignatureConsumer)`
+
+#### System Transactions
+
+The following are new system transactions needed to key rosters and generate ledger signatures.
+
+##### TssMessage System Transaction
+
+TODO: Protobuf of data
+
+##### TssVote System Transaction
+
+TODO: Protobuf of data
+
+##### TssShareSignature System Transaction
+
+TODO: Protobuf of data
+
+#### State Datastructures
+
+The state needs to store the relevant ledger id, consensus rosters, key material, and partially constructed ledger
+signatures.
+
+##### Ledger Id
+
+The `ledgerId` is the public ledger key able to verify ledger signatures. It is used by both the application and the
+platform. This value should not change during address book changes and its value does not change unless the network
+goes through a another TSS bootstrap process. Since its value is not expected to change, storing it in a singleton
+Merkle Leaf by itself is appropriate.
+
+TODO: Protobuf of data
+
+##### Consensus Rosters
+
+This proposal extends the roster data format where each `RosterEntry` has a new long term public EC key called a
+`tssEcKey` that is used in the Groth21 algorithm.
+
+TODO: Protobuf of data
+
+##### Key Material (TSS Data Map)
+
+The TSS Data Map is a Map Merkle Leaf that is a combined map of all the key material for all rosters that are
```

Comment:
I don't think a map is an acceptable leaf type under the upcoming "giant tree" state structure, but my intuitions here are still underdeveloped. Perhaps a synchronous meeting between you, me, Ivan, and whoever else wants to join would be helpful.

For anyone's context, the "giant tree" state structure is the latest plan for how we want to structure our state. There isn't any documentation yet, as this is a new plan handed down from above. I plan to start drafting documentation, but haven't yet had a chance. Timelines also aren't clear to me yet.